### PR TITLE
Remove multiphase from WebUI

### PR DESCRIPTION
**Pull request by squirrelsc**
As requirement of #2390 , removed multiphase from WebUI, document, and put all trials flatten on WebUI.

1. The major change is to split trial-jobs to trials in webui, and use parameter id as part of trial id.
2. If multiphase is enabled, the limitation of trial controls job count only, not trials actually.
3. When multiphase enabled, the trial status may not be right. Previous trials in a job will be marked as success, only last trial presents the job's status.
4. multiphase related documents and UX are removed.

minor changes,
1. Update dev document for webui development.

### CODE REVIEW

File: src/nni_manager/core/nniDataStore.ts [hunk 1]
```diff
@@ -218,20 +218,20 @@ class NNIDataStore implements DataStore {
...
+            const trialInfo: TrialJobInfo | undefined = trialMap.get(key);
+            if (trialInfo === undefined) {
                 continue;
             }
-            if (!(status !== undefined && jobInfo.status !== status)) {
-                if (jobInfo.status === 'SUCCEEDED') {
-                    jobInfo.finalMetricData = finalMetricsMap.get(jobInfo.id);
+            if (!(status !== undefined && trialInfo.status !== status)) {
+                if (trialInfo.status === 'SUCCEEDED') {
+                    trialInfo.finalMetricData = finalMetricsMap.get(trialInfo.id);
```
Comment (QuanluZhang) [inline src/nni_manager/core/nniDataStore.ts:232]: what is the content of `finalMetricData` when there are multiple trials in a job?

Comment (squirrelsc) [reply src/nni_manager/core/nniDataStore.ts:232]: It's all metric data of a trial. No sure why it's called finalMetricData. I just change the variable name to differentiate it from jobMap.

File: docs/en_US/Tutorial/SetupNniDeveloperEnvironment.md [hunk 2]
```diff
@@ -68,7 +68,8 @@ It doesn't need to redeploy, but the nnictl may need to be restarted.
 #### TypeScript
 
 * If `src/nni_manager` will be changed, run `yarn watch` continually under this folder. It will rebuild code instantly.
-* If `src/webui` or `src/nasui` is changed, use **step 3** to rebuild code.
+* If `src/webui` will be changed, run `yarn start` under this folder, and change `MANAGER_IP` like ``//<Web UI IP>:8080/api/v1/nni`` in `src/start/const.ts`. The web UI will refresh automatically if code is changed.
+* If `src/nasui` is changed, use **step 3** to rebuild code.
```
Comment (ultmaster) [inline docs/en_US/Tutorial/SetupNniDeveloperEnvironment.md:72]: If `src/nasui` is changed, use `yarn start` to watch the change of the folder. Meanwhile, `server.js` needs to be launched to read files from `./mockdata`, which contains `graph.json` and other files needed.

File: src/nni_manager/core/nniDataStore.ts (hunk 1 above)
Comment (chicm-ms) [reply src/nni_manager/core/nniDataStore.ts:232]: > what is the content of `finalMetricData` when there are multiple trials in a job?

It is a list of multiple final metrics for each trial

Comment (QuanluZhang) [reply src/nni_manager/core/nniDataStore.ts:232]: I just checked, there are multiple final metrics from mutiple trials. so, `jobInfo` may be a better name.

File: docs/en_US/Tutorial/SetupNniDeveloperEnvironment.md (hunk 2 above)
Comment (squirrelsc) [reply docs/en_US/Tutorial/SetupNniDeveloperEnvironment.md:72]: Update to let user run yarn start. But not understand what user needs to do to launch server.js.

File: src/nni_manager/core/nniDataStore.ts (hunk 1 above)
Comment (squirrelsc) [reply src/nni_manager/core/nniDataStore.ts:232]: After this change, it only get one trial's final metric, not all.

File: src/nni_manager/core/nniDataStore.ts [hunk 3]
```diff
@@ -358,35 +339,94 @@ class NNIDataStore implements DataStore {
...
-                            hParamIds = new Set();
-                        }
-                        if (!hParamIds.has(newHParam.parameter_index)) {
-                            jobInfo.hyperParameters.push(JSON.stringify(newHParam));
-                            hParamIds.add(newHParam.parameter_index);
-                            hParamIdMap.set(jobInfo.id, hParamIds);
+                const hyperParameter: any = this.parseHyperParameter(record.data);
+                if (hyperParameter !== undefined) {
+                    // update trials here.
+                    const trialId = this.getTrialId(jobInfo.id, hyperParameter);
```
Comment (chicm-ms) [inline src/nni_manager/core/nniDataStore.ts:348]: the second parameter name of `getTrialId` is `metricData`,  why pass it a hyperParameter?

Comment (squirrelsc) [reply src/nni_manager/core/nniDataStore.ts:348]: it's called hyperparameters before, I will change the function and variable name to metric data.

File: src/nni_manager/core/nniDataStore.ts [hunk 4]
```diff
@@ -358,35 +339,94 @@ class NNIDataStore implements DataStore {
...
+        let trialId = "";
+        let parameterId = "";
+        if (metricData) {
+            if (metricData.trial_job_id) {
+                trialId = metricData.trial_job_id;
+            } else if (jobId) {
+                trialId = jobId;
+            } else {
+                throw new NNIError("argument null!", "jobId and metricData.trial_job_id shouldn't be undefined both!");
+            }
```
Comment (chicm-ms) [inline src/nni_manager/core/nniDataStore.ts:425]: suggest to have a simpler rule to generate trial id, such as `jobId-parameterId`

File: src/nni_manager/core/nniDataStore.ts [hunk 5]
```diff
@@ -358,35 +339,94 @@ class NNIDataStore implements DataStore {
...
+                    }
+                    trialIds.push(trialId);
                 }
             }
-            if (record.sequenceId !== undefined && jobInfo.sequenceId === undefined) {
-                jobInfo.sequenceId = record.sequenceId;
+        }
+
+        for (const jobTrial of jobTrialMap) {
+            const jobInfo = jobMap.get(jobTrial[0]);
```
Comment (chicm-ms) [inline src/nni_manager/core/nniDataStore.ts:380]: is it possible just expand trials only in webUI  as our discussion?

File: src/nni_manager/core/nniDataStore.ts [hunk 6]
```diff
@@ -358,35 +339,94 @@ class NNIDataStore implements DataStore {
...
-                            hParamIdMap.set(jobInfo.id, hParamIds);
+                const hyperParameter: any = this.parseHyperParameter(record.data);
+                if (hyperParameter !== undefined) {
+                    // update trials here.
+                    const trialId = this.getTrialId(jobInfo.id, hyperParameter);
+                    let trialInfo: TrialJobInfo | undefined;
+                    if (trialMap.has(trialId)) {
+                        trialInfo = trialMap.get(trialId);
+                        if (trialInfo && trialInfo.hyperParameters) {
+                            trialInfo.hyperParameters.push(JSON.stringify(hyperParameter));
```
Comment (QuanluZhang) [inline src/nni_manager/core/nniDataStore.ts]: there are mutiple `hyperParameter` for a `trialInfo` whose `trialId` is "jobid-parameterid"?

File: src/nni_manager/core/nniDataStore.ts [hunk 7]
```diff
@@ -358,35 +339,94 @@ class NNIDataStore implements DataStore {
...
-                    if (jobInfo.hyperParameters !== undefined) {
-                        let hParamIds: Set<number> | undefined = hParamIdMap.get(jobInfo.id);
-                        if (hParamIds === undefined) {
-                            hParamIds = new Set();
-                        }
-                        if (!hParamIds.has(newHParam.parameter_index)) {
-                            jobInfo.hyperParameters.push(JSON.stringify(newHParam));
-                            hParamIds.add(newHParam.parameter_index);
-                            hParamIdMap.set(jobInfo.id, hParamIds);
+                const metricData: any = this.parseMetricData(record.data);
```
Comment (chicm-ms) [inline src/nni_manager/core/nniDataStore.ts:345]: the `record.data` here is actually a hyperParameter stored using `storeTrialJobEvent`

File: src/nni_manager/core/nniDataStore.ts [hunk 8]
```diff
@@ -358,35 +339,94 @@ class NNIDataStore implements DataStore {
...
+                    if (trialMap.has(trialId)) {
+                        trialInfo = trialMap.get(trialId);
+                        if (trialInfo && trialInfo.hyperParameters) {
+                            trialInfo.hyperParameters.push(JSON.stringify(hyperParameter));
                         }
                     } else {
-                        assert(false, 'jobInfo.hyperParameters is undefined');
+                        trialInfo = {
+                            id: trialId,
+                            status: this.getJobStatusByLatestEvent('RUNNING', record.event),
```
Comment (QuanluZhang) [inline src/nni_manager/core/nniDataStore.ts:358]: why the default status is 'RUNNING'?

File: src/nni_manager/core/nniDataStore.ts [hunk 9]
```diff
@@ -358,35 +339,94 @@ class NNIDataStore implements DataStore {
...
+                            hyperParameters: [JSON.stringify(hyperParameter)],
+                            startTime: record.timestamp,
+                            sequenceId: sequenceId++
```
Comment (QuanluZhang) [inline src/nni_manager/core/nniDataStore.ts:361]: `sequenceId` is generated by this function not from the record?

File: src/nni_manager/core/nniDataStore.ts [hunk 10]
```diff
@@ -358,35 +339,94 @@ class NNIDataStore implements DataStore {
...
+                if (!trialInfo) {
+                    throw new Error('Empty trialInfo');
+                }
+                if (lastTrialInfo) {
+                    lastTrialInfo.endTime = trialInfo.startTime;
+                }
+                trialInfo.logPath = jobInfo.logPath;
+                // There is no status reported for each trial. So assume it's success, if it's not last one.
+                // The last trial status is the same as job.
+                trialInfo.status = 'SUCCEEDED';
```
Comment (QuanluZhang) [inline src/nni_manager/core/nniDataStore.ts:400]: what if a parameter does not have a matched final result, due to catched error by user?

File: src/nni_manager/core/sqlDatabase.ts [hunk 11]
```diff
@@ -178,17 +178,21 @@ class SqlDB implements Database {
     public queryTrialJobEvent(trialJobId?: string, event?: TrialJobEvent): Promise<TrialJobEventRecord[]> {
         let sql: string = '';
         let args: any[] | undefined;
+        let jobId: string | undefined;
+        if (trialJobId && trialJobId.search("-") > 0) {
```
Comment (QuanluZhang) [inline src/nni_manager/core/sqlDatabase.ts:182]: `trialJobId` is in "jobid-parameterid" format?

File: src/nni_manager/core/sqlDatabase.ts [hunk 12]
```diff
@@ -215,17 +219,21 @@ class SqlDB implements Database {
...
         if (metricType === undefined && trialJobId === undefined) {
             sql = 'select * from MetricData';
         } else if (trialJobId === undefined) {
             sql = 'select * from MetricData where type=?';
             args = [metricType];
         } else if (metricType === undefined) {
-            sql = 'select * from MetricData where trialJobId=?';
-            args = [trialJobId];
+            sql = 'select * from MetricData where trialJobId=? and parameterId=?';
+            args = [jobId, parameterId];
```
Comment (QuanluZhang) [inline src/nni_manager/core/sqlDatabase.ts:233]: `parameterId` is in data field, how to query `parameterId`?

File: src/nni_manager/core/nniDataStore.ts (hunk 7 above)
Comment (squirrelsc) [reply src/nni_manager/core/nniDataStore.ts:345]: the data is as below, it called metric data, not only hyper parameters.

{"parameter_id": 2, "parameter_source": "algorithm", "parameters": {"batch_size": 128, "hidden_size": 512, "lr": 0.01, "momentum": 0.738026583665898}, "parameter_index": 0}
--

File: src/nni_manager/core/nniDataStore.ts (hunk 8 above)
Comment (squirrelsc) [reply src/nni_manager/core/nniDataStore.ts:358]: it requested parameter if there is record, so it's considered as running.

File: src/nni_manager/core/nniDataStore.ts (hunk 9 above)
Comment (squirrelsc) [reply src/nni_manager/core/nniDataStore.ts]: It's not sequenceId of metric record, it's trial number shows on UI. It has value for first request only, for following are empty. so it needs to reassign value here. I changed the variable name to trialNumber for clearer.

File: src/nni_manager/core/nniDataStore.ts (hunk 10 above)
Comment (squirrelsc) [reply src/nni_manager/core/nniDataStore.ts:400]: As it needs metric data record to parse this kind of problem, and it's high performance impact in this function.

File: src/nni_manager/core/nniDataStore.ts (hunk 4 above)
Comment (squirrelsc) [reply src/nni_manager/core/nniDataStore.ts]: updated

File: src/nni_manager/core/nniDataStore.ts (hunk 6 above)
Comment (squirrelsc) [reply src/nni_manager/core/nniDataStore.ts]: no, there should be only one, the design is not changed this time.

File: src/nni_manager/core/sqlDatabase.ts (hunk 11 above)
Comment (squirrelsc) [reply src/nni_manager/core/sqlDatabase.ts:182]: yes

File: src/nni_manager/core/sqlDatabase.ts (hunk 12 above)
Comment (squirrelsc) [reply src/nni_manager/core/sqlDatabase.ts:233]: MetricData table has this field also.

# | timestamp | trialJobId | parameterId | type | sequence | data
-- | -- | -- | -- | -- | -- | --
1 | 1588846040502 | fWpr7 | 2 | PERIODICAL | 0 | "97.41"

File: src/nni_manager/core/nniDataStore.ts (hunk 7 above)
Comment (chicm-ms) [reply src/nni_manager/core/nniDataStore.ts:345]: the above is a hyper parameter,  the content of parameters is generated by Tuner and other fields are meta-data,   set by nni sdk.

Comment (squirrelsc) [reply src/nni_manager/core/nniDataStore.ts:345]: Yes, but it was called as metricData... I didn't change it. And the actual metric data called MetricDataRecord.

File: src/nni_manager/core/nniDataStore.ts [hunk 13]
```diff
@@ -242,16 +242,10 @@ class NNIDataStore implements DataStore {
...
-
         for (const metric of metrics) {
             const existMetrics: MetricDataRecord[] | undefined = map.get(metric.trialJobId);
             if (existMetrics !== undefined) {
-                if (!multiPhase) {
-                    this.log.error(`Found multiple FINAL results for trial job ${trialJobId}, metrics: ${JSON.stringify(metrics)}`);
-                } else {
-                    existMetrics.push(metric);
-                }
+                existMetrics.push(metric);
```
Comment (SparkSnail) [inline src/nni_manager/core/nniDataStore.ts:248]: Seems it need to contain the content from `if` condition, not `else`.

Comment (squirrelsc) [reply src/nni_manager/core/nniDataStore.ts]: The change is reverted, there is no change on nniDataStore now.

**Pull request merged**